### PR TITLE
[WIP] New model : MAE

### PR DESCRIPTION
**Pull request by skpig**
Adding new model MAE in image_classification.
Paper: Masked Autoencoders Are Scalable Vision Learners
arxiv: https://arxiv.org/abs/2111.06377

### CODE REVIEW

**Comment by skpig** [comment]
@xperzy  朱老师，我已经把MAE的pretraining部分写完了，您如果最近有时间可以review一下代码，然后跑一下。
有几个地方的代码我还存在疑问，需要您注意一下：
1. 论文中提出一种通过shuffle进行mask的方法，为了提高算法效率，我对同一个batch内的所有image做了完全相同的shuffle，也即做了完全相同的mask。
2. cls token是否应该加上对应的positional embedding
3. 因为不清楚您能调用的显卡资源，batch_size，accum_iter的大小需要您自己按实际情况设置，totalbatchsize我都已经写在对应的.yaml文件中
https://github.com/BR-IDL/PaddleViT/blob/370327bf68208e0b5607240ddf6beae1c7f17988/image_classification/MAE/configs/vit_base_patch16_224_pretrain.yaml#L28
4. 论文中汇报的结果都是基于ViT-L/16，如果可能，还是尽量训练ViT-Large model
5. 我始终无法通过`test_all_parameters_updated`测试，所有的参数梯度都为0，但奇怪的是，我尝试训练了MAE，损失正常下降，如果方便的话麻烦您帮我检查一下原因。
https://github.com/BR-IDL/PaddleViT/blob/91d3ca966bbaab5fd241bbc8208047f05327b0e7/image_classification/MAE/tests/test_transformer.py#L48-L58

**Comment by xperzy** [comment]
Good Work! 

There are several part we can discuss and fix:

1.  nn.Layer
https://github.com/BR-IDL/PaddleViT/blob/deccdad4bfe1d51277f7297c485ed64c470dfb7e/image_classification/MAE/transformer.py#L328

This layer may not need to be a nn.Layer, since it doesn't have any learnable paramter, the `mask_token` is actually a member of its upper class `MAEPretrinTransformer`, I think we can modify this as methods instead of nn.Layer.


2.  variable name
https://github.com/BR-IDL/PaddleViT/blob/deccdad4bfe1d51277f7297c485ed64c470dfb7e/image_classification/MAE/transformer.py#L622
Such as 'image', 'mask image', 'x', etc, we can use more clear var names, which will make the code more readable.

I would suggest refer to the original paper and follow our paddlevit name style.  BTW, shall we consider the 'label' here may be renamed to 'target' or 'target_patch'? 

3.  效率问题
 可以尝试分析下，图像块加载的位置是否对性能有影响， Option1. 当前的加载方式， Option 2. 数据切块在dataloader/dataset里做。

**Comment by xperzy** [comment]
Please pass the codefactor check also. Thanks

**Comment by xperzy** [comment]
@skpig 代码中的模型初始化部分建议修改成DeiT的方式，当前的ViT权重初始化可能对训练有一定的影响

**Comment by skpig** [comment]
@xperzy 我在按照SwinTransformer更改main_multi_gpu.py的时候遇到了如下代码段，不太明白它的含义，特别是其中`512.0`，麻烦朱老师帮我解答一下。
https://github.com/BR-IDL/PaddleViT/blob/21826cf51999d1b4836faf5ed0a68e7c8d9dff49/image_classification/SwinTransformer/main_multi_gpu.py#L356-L371

**Comment by SigureMo** [comment]
> @xperzy 我在按照SwinTransformer更改main_multi_gpu.py的时候遇到了如下代码段，不太明白它的含义，特别是其中`512.0`，麻烦朱老师帮我解答一下。
> 
> https://github.com/BR-IDL/PaddleViT/blob/21826cf51999d1b4836faf5ed0a68e7c8d9dff49/image_classification/SwinTransformer/main_multi_gpu.py#L356-L371

imho，这里 512 应当是指默认配置里的 learning rate （5e-4）是按照 batch size 为 512 时设置的（论文中 bs=1024 时 lr=1e-3），但随着 batch size 的增大，lr 是应当自适应缩放的，这里其实一共有三次缩放：

- 第一次缩放是根据你所设置的 batch size 线性缩放，`lr *= batch_size / batch_size_default`
- 第二次是根据使用多卡的节点数进行缩放，这相当于提升了 batch size，`lr *= world_size`
- 第三次是根据几个 step 进行一次 gradient descent 进行缩放，因为多个 step 后 GD 相当于变相提高 batch size，`lr *= accum_iter`

不过另外一个版本的代码（貌似是更新的版本）中貌似没有考虑 world size 的影响，这里的 LINEAR_SCALED_LR 我的理解是上面所说的第一次缩放中的 `batch_size_default`，但我不明白这里为什么少了第二次根据 world_size 缩放 lr 这一步（在该项目其余代码中也没找到相关逻辑），这里可以麻烦朱老师解答一下嘛～

https://github.com/BR-IDL/PaddleViT/blob/21826cf51999d1b4836faf5ed0a68e7c8d9dff49/image_classification/ConvMixer/main_multi_gpu.py#L355-L372

**Comment by skpig** [comment]
@SigureMo 感谢解答！这种lr随batch size等比例增大是通用且常见的吗？还是说也仅仅是该论文里一个方便的小trick，具体lr还是要根据训练情况来调整呢？

**Comment by SigureMo** [comment]
@skpig  具体的细节我也不太清楚，你可以听一下朱老师在 DeiT 课程里讲的（不知道你有木有参与那门课程～），按照老师在课程里说的貌似 CNN 中就有这种 trick 了，DeiT 中沿用了该 trick，在之后其余 ViT 网络基本都是参考 DeiT 的一些 trick 进行训练的（个人理解，建议你看一下视频这一部分）

https://aistudio.baidu.com/aistudio/education/lessonvideo/2012333 中 26:33 秒左右

**Comment by skpig** [comment]
> @skpig 具体的细节我也不太清楚，你可以听一下朱老师在 ViT 课程 DeiT 章节里讲的（不知道你有木有参与那门课程～），按照老师在课程里说的貌似 CNN 中就有这种 trick 了，DeiT 中沿用了该 trick，在之后其余 ViT 网络基本都是参考 DeiT 的一些 trick 进行训练的（个人理解，建议你看一下视频这一部分）
> 
> https://aistudio.baidu.com/aistudio/education/lessonvideo/2012333 中 26:33 秒左右
@SigureMo 十分感谢！

@xperzy 。朱老师，MAE finetune部分的代码也已经完成。另外，MAE finetune使用了RandAugment，当前仓库里我没有看到对应模块代码，所以就自己写了一个，麻烦朱老师review一下。

**Review by lmk123568 — COMMENTED** [review]
please delete

* .idea/deployment.xml
* .idea/inspectionProfiles/Project_Default.xml

thank you for your work

**Comment by skpig** [comment]
Sorry for my mistakes. 
Is the module RandAugment correct ? Thanks for your review. 
And if everything works fine, maybe we can do some experiments following MAE paper to test the performance/correctness of my code.

**Comment by xperzy** [comment]
@skpig 我发现有两个branch： `MAE` 和 `mae`,  我最近的更新是在`MAE` branch下的， 其中的main 函数好像也是比较新的版本，但是当前`mae`branch下也有randaug之类的更新，这两个branch能否合并一下？

**Comment by xperzy** [comment]
@skpig 我发现有两个branch： `MAE` 和 `mae`,  我最近的更新是在`MAE` branch下的， 其中的main 函数好像也是比较新的版本，但是当前`mae`branch下也有randaug之类的更新，这两个branch能否合并一下？

**Comment by xperzy** [comment]
I  have refactor several code and now this `mae` branch is our updated MAE version.

**Comment by xperzy** [comment]
Current status, multi gpu version is running ok.

```
INFO:local_logger:Now training epoch 1. LR=0.000023
INFO:local_logger:----- Total # of train batch (single gpu): 626
INFO:local_logger:Start training from epoch 1.
INFO:local_logger:Now training epoch 1. LR=0.000023
INFO:local_logger:----- Total # of train batch (single gpu): 626
INFO:local_logger:Start training from epoch 1.
INFO:local_logger:----- Total # of train batch (single gpu): 626
INFO:local_logger:Now training epoch 1. LR=0.000023
INFO:local_logger:Start training from epoch 1.
INFO:local_logger:Now training epoch 1. LR=0.000023
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5893
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5086
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5884
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.6202
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5317
INFO:master_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5601
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5260
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.5154
INFO:local_logger:Epoch[001/800], Step[0000/0626], Avg Loss: 1.6014
```

File: image_classification/MAE/main_multi_gpu_finetune.py [hunk 1]
```diff
@@ -27,14 +27,13 @@
 import paddle.distributed as dist
 from datasets import get_dataloader
 from datasets import get_dataset
-from transformer import build_mae_finetune as build_model
+from transformer import build_mae_pretrain as build_model
```
Comment (skpig) [inline image_classification/MAE/main_multi_gpu_finetune.py:30]: This file is used to finetune model. Why use `build_mae_pretrain` to build model instead of `build_mae_finetune`

File: image_classification/MAE/configs/vit_base_patch16_224_pretrain_dec1.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            EMBED_DIM: 512
+            DEPTH: 1
+            NUM_HEADS: 8
+TRAIN:
+    NUM_EPOCHS: 800
+    WARMUP_EPOCHS: 40
+    WEIGHT_DECAY: 0.05
+    BASE_LR: 0.75e-4 #1.5e-4
+    WARMUP_START_LR: 1e-6
+    GRAD_CLIP: 5.0
```
Comment (skpig) [inline image_classification/MAE/configs/vit_base_patch16_224_pretrain_dec1.yaml:27]: Just to remind you, the original paper doesn't use gradient clip.

File: image_classification/MAE/main_multi_gpu_finetune.py (hunk 1 above)
Comment (xperzy) [reply image_classification/MAE/main_multi_gpu_finetune.py]: This is now fixed, should use finetune.

File: image_classification/MAE/configs/vit_base_patch16_224_pretrain_dec1.yaml (hunk 2 above)
Comment (xperzy) [reply image_classification/MAE/configs/vit_base_patch16_224_pretrain_dec1.yaml:27]: Now the training still have NaN error after dozens of epochs, still trying these hyperparams.

**Review by skpig — COMMENTED** [review]
The codes about masking is fantastic !!!

**Comment by xperzy** [comment]
Now the code seems OK while training on 8 GPUs,  the following log shows it runs OK until 260+ epochs, therefore I'm merging this repo. Note that the hyperparams may need to change if training on larger num of gpus, I will update the code and pre-trained models once the training is done.


```
2021-12-22 10:51:14,252 Now training epoch 268. LR=0.000222
2021-12-22 10:51:29,584 Epoch[268/800], Step[0000/0626], Avg Loss: 0.6928
2021-12-22 10:52:46,924 Epoch[268/800], Step[0100/0626], Avg Loss: 0.6913
2021-12-22 10:54:07,803 Epoch[268/800], Step[0200/0626], Avg Loss: 0.6915
2021-12-22 10:55:25,311 Epoch[268/800], Step[0300/0626], Avg Loss: 0.6915
2021-12-22 10:56:42,835 Epoch[268/800], Step[0400/0626], Avg Loss: 0.6913
2021-12-22 10:58:01,682 Epoch[268/800], Step[0500/0626], Avg Loss: 0.6914
2021-12-22 10:59:21,999 Epoch[268/800], Step[0600/0626], Avg Loss: 0.6914
2021-12-22 10:59:45,334 ----- Epoch[268/800], Train Loss: 0.6914, time: 511.08
2021-12-22 10:59:49,043 ----- Save model: ./output/train-20211220-18-15-57/MAE-Epoch-268-Loss-0.69200028
2021-12-22 10:59:49,043 ----- Save optim: ./output/train-20211220-18-15-57/MAE-Epoch-268-Loss-0.69200028
2021-12-22 10:59:49,043 Now training epoch 269. LR=0.000223
2021-12-22 11:00:05,258 Epoch[269/800], Step[0000/0626], Avg Loss: 0.6902
2021-12-22 11:01:27,665 Epoch[269/800], Step[0100/0626], Avg Loss: 0.6910
2021-12-22 11:02:50,972 Epoch[269/800], Step[0200/0626], Avg Loss: 0.6912
2021-12-22 11:04:16,962 Epoch[269/800], Step[0300/0626], Avg Loss: 0.6912
2021-12-22 11:05:42,827 Epoch[269/800], Step[0400/0626], Avg Loss: 0.6913
2021-12-22 11:07:09,037 Epoch[269/800], Step[0500/0626], Avg Loss: 0.6913
2021-12-22 11:08:37,468 Epoch[269/800], Step[0600/0626], Avg Loss: 0.6913
2021-12-22 11:09:02,119 ----- Epoch[269/800], Train Loss: 0.6913, time: 553.07
2021-12-22 11:09:06,060 ----- Save model: ./output/train-20211220-18-15-57/MAE-Epoch-269-Loss-0.69151166
2021-12-22 11:09:06,060 ----- Save optim: ./output/train-20211220-18-15-57/MAE-Epoch-269-Loss-0.69151166
2021-12-22 11:09:06,060 Now training epoch 270. LR=0.000223
2021-12-22 11:09:23,607 Epoch[270/800], Step[0000/0626], Avg Loss: 0.6945
2021-12-22 11:10:48,461 Epoch[270/800], Step[0100/0626], Avg Loss: 0.6917
2021-12-22 11:12:15,911 Epoch[270/800], Step[0200/0626], Avg Loss: 0.6914
2021-12-22 11:13:42,944 Epoch[270/800], Step[0300/0626], Avg Loss: 0.6913
2021-12-22 11:15:08,887 Epoch[270/800], Step[0400/0626], Avg Loss: 0.6913
2021-12-22 11:16:36,558 Epoch[270/800], Step[0500/0626], Avg Loss: 0.6913
2021-12-22 11:18:03,601 Epoch[270/800], Step[0600/0626], Avg Loss: 0.6913
```
